operator/pkg/bgpv2/cluster.go

```diff
@@ -76,13 +68,133 @@ func (b *BGPResourceManager) reconcileBGPClusterConfig(ctx context.Context, conf
 
 	// Call API only when there's a condition change
 	if updateStatus {
-		_, uErr := b.clientset.CiliumV2alpha1().CiliumBGPClusterConfigs().UpdateStatus(ctx, config, meta_v1.UpdateOptions{})
-		if uErr != nil {
-			err = errors.Join(err, uErr)
+		_, err := b.clientset.CiliumV2alpha1().CiliumBGPClusterConfigs().UpdateStatus(ctx, config, meta_v1.UpdateOptions{})
+		if err != nil {
+			err = errors.Join(errs, err)
 		}
 	}
 
-	return err
+	return errs
+}
+
+func (b *BGPResourceManager) upsertNodeConfigs(ctx context.Context, config *v2alpha1.CiliumBGPClusterConfig) (sets.Set[string], sets.Set[string], error) {
+	var nodeSelector slim_labels.Selector
+	if config.Spec.NodeSelector == nil {
+		// nil selector means select all nodes
+		nodeSelector = slim_labels.Everything()
+	} else {
+		selector, err := slim_meta_v1.LabelSelectorAsSelector(config.Spec.NodeSelector)
+		if err != nil {
+			return nil, nil, err
+		}
+		nodeSelector = selector
+	}
+
+	// Name of the nodes selected by nodeSelector
+	matchingNodes := sets.New[string]()
+
+	// Name of the ClusterConfig selecting the same node
+	conflictingClusterConfigs := sets.New[string]()
+
+	// Errors
+	var errs error
+
+	for _, node := range b.ciliumNodeStore.List() {
+		if !nodeSelector.Matches(slim_labels.Set(node.Labels)) {
+			continue
+		}
+
+		// Record selected node for later use
+		matchingNodes.Insert(node.Name)
+
+		// Find node config for this node
+		oldNodeConfig, oldNodeConfigExists, err := b.nodeConfigStore.GetByKey(resource.Key{Name: node.Name})
+		if err != nil {
+			errs = errors.Join(errs, err)
+			continue
+		}
+
+		// Conflict detection
+		if oldNodeConfigExists && !isOwner(oldNodeConfig.OwnerReferences, config) {
+			owner := ownerClusterConfigName(oldNodeConfig.OwnerReferences)
+			if owner != "" {
+				conflictingClusterConfigs.Insert(owner)
+			}
+			// Conflict detected. Skip this node.
+			continue
+		}
+
+		// Find node overrides for this node
+		nodeConfigOverride, nodeConfigOverrideExists, err := b.nodeConfigOverrideStore.GetByKey(resource.Key{Name: node.Name})
+		if err != nil {
+			errs = errors.Join(errs, err)
+			continue
+		}
+
+		// Build a desired node config
+		var overrideInstances []v2alpha1.CiliumBGPNodeConfigInstanceOverride
+		if nodeConfigOverrideExists {
+			overrideInstances = nodeConfigOverride.Spec.BGPInstances
+		}
+		newNodeConfig := &v2alpha1.CiliumBGPNodeConfig{
+			ObjectMeta: meta_v1.ObjectMeta{
+				Name: node.Name,
+				OwnerReferences: []meta_v1.OwnerReference{
+					{
+						APIVersion: v2alpha1.SchemeGroupVersion.String(),
+						Kind:       v2alpha1.BGPCCKindDefinition,
+						Name:       config.GetName(),
+						UID:        config.GetUID(),
+						Controller: ptr.To(true),
```

Comment:
Any reason to add Controller = true?

Reply:
I should have a comment on this, sorry. It's generally recommended to set it to indicate the resource is "the" controller of the resource. The API server guarantees that the only one owner sets the owner reference with the `controller` field (the controller came later get and error like `Only one reference can have Controller set to true`). We do our own conflict detection, so it's not absolutely necessary tho.